client_request.js

```diff
@@ -0,0 +1,208 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+'use strict';
+
+var ClientRequestErrors = require('./client_request_errors.js');
+var ClientRequestStates = require('./client_request_states.js');
+var EventEmitter = require('events').EventEmitter;
+var LoggingLevels = require('./lib/logging/levels.js');
+var safeParse = require('./lib/util.js').safeParse;
+var validateHostPort = require('./lib/util.js').validateHostPort;
+var uuid = require('node-uuid');
+var util = require('util');
+
+/* jshint maxparams: 6 */
+function ClientRequest(client, opts, endpoint, head, body, callback) {
+    this.client = client;
+    this.opts = opts;
+    this.endpoint = endpoint;
+    this.head = head;
+    this.body = body;
+    this.callback = callback;
+    this.id = uuid.v4();
+    this.timestamp = Date.now();
+    this.isCanceled = false;
+    this.state = ClientRequestStates.initialized;
+}
+
+util.inherits(ClientRequest, EventEmitter);
+
+ClientRequest.prototype.cancel = function cancel() {
+    this.cancelTimestamp = Date.now();
+    this.isCanceled = true;
+    this._invokeCallback(ClientRequestErrors.RequestCanceledError({
+        endpoint: this.endpoint
+    }));
+};
+
+ClientRequest.prototype.send = function send() {
+    var self = this;
+    var opts = this.opts;
+    var endpoint = this.endpoint;
+    var timeout = opts.timeout || 30000;
+    var subChannel = this.client.subChannel;
+
+    if (subChannel && subChannel.destroyed) {
+        process.nextTick(function onTick() {
+            self._invokeCallback(ClientRequestErrors.ChannelDestroyedError({
+                endpoint: endpoint,
+                channelType: 'subChannel'
+            }));
+        });
+        return;
+    }
+
+    if (subChannel &&
+            subChannel.topChannel &&
+            subChannel.topChannel.destroyed) {
```

Comment:
question: can a topchannel be destroyed, without the subchannel having also been destroyed? ie, does the ordering here make it impossible for this 2nd one to ever fire?